[New Integration] blacklens.io

ChriZzn opened the pull request:
## Description

[blacklens.io](https://blacklens.io) is a powerful Attack Surface Management (ASM) platform designed to help organizations monitor and secure their external digital footprint. It provides in-depth insights into potential threats, including vulnerabilities and darknet activities, giving you a comprehensive view of your attack surface.

## Architecture

This is an API-driven integration. Refer to the Knowledge Base for a full list of available API calls.

## Integration release checklist
This checklist is intended for integrations maintainers to ensure consistency
when creating or updating a Package, Module or Dataset for an Integration.

- [x] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/main/CONTRIBUTING.md)
- [x] Supported versions of the monitoring target are documented
- [x] Supported operating systems are documented (if applicable)
- [x] Integration or System tests exist
- [x] Documentation exists, useful guidelines to follow
- [x] Fields follow ECS and naming conventions
- [x] At least a manual test with ES / Kibana / Agent has been performed.
- [x] Required Kibana version set to: 8.15.2

## Screenshots

![Screenshot 2024-11-12 at 11 12 43](https://github.com/user-attachments/assets/e8f19566-af29-4f41-b8f5-f9a14de99228)

![blacklens-dashboard-alerts](https://github.com/user-attachments/assets/900a39ad-7df0-4111-a342-6b88f081b9d4)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18198)
* Commit: 89c0eb002b61c9628b0d10b531775f2212b71084

### Failed CI Steps
* [:white_check_mark: Check go sources](https://buildkite.com/elastic/integrations/builds/18198#01931f00-16a4-42dd-9f12-f1fbab300716)

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18198","number":18198,"commit":"89c0eb002b61c9628b0d10b531775f2212b71084"}],"number":18198}
buildkite-pr-comment-->

kcreddy commented on packages/blacklens/docs/README.md:
```diff
@@ -0,0 +1,27 @@
+3. Go to **Profile → Generate API Key** and copy it.
+6. Go to **Settings → Company**.
+7. Copy **ws_id** and **tenant_id**.
+
```
After copying the values, can you add a section to setup the integration in Elastic? This helps users understand which values should be copied into respective fields.
[Example](https://www.elastic.co/guide/en/integrations-developer/current/documentation-guidelines.html)

kcreddy replied:
Also document here if the users need any additional permissions/privileges to read the Alerts?

kcreddy commented on packages/blacklens/docs/README.md:
```diff
@@ -0,0 +1,27 @@
+6. Go to **Settings → Company**.
+7. Copy **ws_id** and **tenant_id**.
+
+For detailed setup instructions, please refer to the blacklens.io Knowledge Base.
```
Can you add a link to `blacklens.io Knowledge Base`?

ChriZzn replied:
The Knowledge Base is for registered users only and is located within the portal.

kcreddy commented on packages/blacklens/docs/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Setup
+
+1. Login to your blacklens.io Portal (This URL will be used for the Instance URL)
```
If it has a generic portal URL, please add it. Looks like it is a SaaS platform, so probably has one.

kcreddy commented on packages/blacklens/docs/README.md:
```diff
@@ -0,0 +1,27 @@
+
+The blacklens.io integration collects one type of data streams: logs
+
+**Alerts** returns a list of blacklens.io alerts
```
Add a link to the Alerts API documentation.
[Example](https://github.com/elastic/integrations/blob/main/packages/abnormal_security/_dev/build/docs/README.md?plain=1#L11)

ChriZzn replied:
The link is referenced within the portal

kcreddy commented on packages/blacklens/data_stream/alerts/agent/stream/httpjson.yml.hbs:
```diff
@@ -0,0 +1,63 @@
+{{#if ssl}}
+request.ssl: {{ssl}}
+{{/if}}
+
```
Also add:
```yaml
{{#if http_client_timeout}}
request.timeout: {{http_client_timeout}}
{{/if}}
```

kcreddy commented on packages/blacklens/data_stream/alerts/agent/stream/httpjson.yml.hbs:
```diff
@@ -0,0 +1,63 @@
+response.split:
+  target: body.results
+  ignore_empty_value: true
+
+
```
```suggestion

```

kcreddy commented on packages/blacklens/data_stream/alerts/agent/stream/httpjson.yml.hbs:
```diff
@@ -0,0 +1,63 @@
+{{#each tags as |tag|}}
+  - {{tag}}
+{{/each}}
+
```
Also add
```yaml
{{#contains "forwarded" tags}}
publisher_pipeline.disable_host: true
{{/contains}}

[Example](https://github.com/elastic/integrations/blob/main/packages/1password/data_stream/audit_events/agent/stream/httpjson.yml.hbs#L56-L58)
```

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,109 @@
+      field: message
+      ignore_missing: true
+      if: 'ctx.event?.original != null'
+  - set:
+      field: event.kind
+      value: alert
```
For space efficiency, for constants like these, instead of adding them to ingest pipeline, can you add directly to the fields as `constant_keyword`?

Example: https://github.com/elastic/integrations/blob/main/packages/github/data_stream/code_scanning/fields/ecs.yml

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,109 @@
+  - set:
+      field: event.kind
+      value: alert
+  - set:
+      field: event.ingested
+      value: '{{_ingest.timestamp}}'
```
This is automatically set by the fleet's pipeline which always runs after this main pipeline. You can remove this.

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,109 @@
+  - set:
+      field: event.ingested
+      value: '{{_ingest.timestamp}}'
+  - set:
+      field: event.category
+      value: [threat]
```
```suggestion
  - append:
      field: event.category
      value: threat
      allow_duplicates: false
```

kcreddy replied:
Same for below

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,109 @@
+        - json.id
+      target_field: _id
+      ignore_missing: true
+  - date:
+      field: json.created_date
+      formats:
+        - ISO8601
+  - date:
+      field: json.updated_date
+      target_field: blacklens.alert.updated_date
+      formats:
+        - ISO8601
```
This will make all updates for `alert` to have  same `@timestamp` value. Is this intended?

kcreddy replied:
Also, your current fingerprint only includes `json.id`. 
So, any updates to the `alert` is also not ingested.
https://github.com/elastic/integrations/blob/89c0eb002b61c9628b0d10b531775f2212b71084/packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml#L31-L35

If you would like to ingest updates to an alert, you should also add `json.updated_date` to the fingerprint processor.

ChriZzn replied:
Hi, yes to have the same @timestamp regardless of updates, I added json.updated_date to the fingerprint function.

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,109 @@
+      field: json
+      ignore_failure: true
+      ignore_missing: true
+  - remove:
+      field: event.original
+      ignore_missing: true
+      ignore_failure: true
+      if: ctx.tags == null || !(ctx.tags.contains('preserve_original_event'))
```
Automatically handled in fleet's pipeline. This can be removed.

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,109 @@
+      ignore_missing: true
+  - remove:
+      field: json
+      ignore_failure: true
```
```suggestion
```

kcreddy commented on packages/blacklens/docs/README.md:
```diff
@@ -0,0 +1,27 @@
+6. Go to **Settings → Company**.
+7. Copy **ws_id** and **tenant_id**.
+
+For detailed setup instructions, please refer to the blacklens.io Knowledge Base.
```
At the end of the file, please also add fields and sample event reference. [Example](https://github.com/elastic/integrations/blob/main/packages/abnormal_security/_dev/build/docs/README.md?plain=1#L65-L75)

This will be rendered inside `docs/README.md`

kcreddy commented on packages/blacklens/img/blacklens-dashboard.png:
Can you rename this `blacklens-dashboards-alerts.png`?
This should help with other data points down the line.

kcreddy commented on packages/blacklens/manifest.yml:
```diff
@@ -0,0 +1,100 @@
+  elastic:
+    subscription: "basic"
+screenshots:
+  - src: /img/blacklens-dashboard.png
```
```suggestion
  - src: /img/blacklens-dashboard-alerts.png
```

Rename the file accordingly.

kcreddy commented on packages/blacklens/manifest.yml:
```diff
@@ -0,0 +1,100 @@
+    subscription: "basic"
+screenshots:
+  - src: /img/blacklens-dashboard.png
+    title: blacklens.io Dashboard
```
```suggestion
    title: blacklens.io Alerts Dashboard
```

kcreddy commented on packages/blacklens/manifest.yml:
```diff
@@ -0,0 +1,100 @@
+            description: Workspace ID of blacklens.io.
+            multi: false
+            required: true
+            show_user: true
```
In the `README`, I see a mention to copy the `tenant_id`, but it is not being used here.

kcreddy commented on packages/blacklens/manifest.yml:
```diff
@@ -0,0 +1,100 @@
+              #    sxSmbIUfc2SGJGCJD4I=
+              #    -----END CERTIFICATE-----
+owner:
+  github: elastic/integrations
```
```suggestion
  github: elastic/security-service-integrations
```

kcreddy replied:
Also add this line to `.github/CODEOWNERS` file:
```yaml
/packages/blacklens @elastic/security-service-integrations
```
[Example](https://github.com/elastic/integrations/blob/main/.github/CODEOWNERS#L121)

kcreddy commented:
@ChriZzn , thanks for the contribution 🥇 
The CI is failing on missing codeowners. Once [this](https://github.com/elastic/integrations/pull/11664#discussion_r1837618218) is added, this error should be fixed.

Also, is it possible to add pipeline tests with few sample events? [Example](https://github.com/elastic/integrations/tree/main/packages/abnormal_security/data_stream/ai_security_mailbox/_dev/test/pipeline)

Reference for adding pipeline tests: https://www.elastic.co/guide/en/integrations-developer/current/pipeline-testing.html

ChriZzn commented:
> @ChriZzn , thanks for the contribution 🥇
> 
> 1. The CI is failing on missing codeowners. Once [this](https://github.com/elastic/integrations/pull/11664#discussion_r1837618218) is added, this error should be fixed.
> 2. Also, is it possible to add pipeline tests with few sample events? [Example](https://github.com/elastic/integrations/tree/main/packages/abnormal_security/data_stream/ai_security_mailbox/_dev/test/pipeline)
>    Reference for adding pipeline tests: https://www.elastic.co/guide/en/integrations-developer/current/pipeline-testing.html
> 3. Could you add few screenshots from your integration and dashboard(s) to your PR in the **Screenshots** section?
>    [Example](https://github.com/elastic/integrations/pull/10617#issue-2432282642)

Hi i fixed and added all requested items... Regards Christoph

kcreddy commented:
/test

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18267)
* Commit: 582da2f731faeb5f9e521fd88dd7233fbe8f597a

### Failed CI Steps
* [Check integrations blacklens](https://buildkite.com/elastic/integrations/builds/18267#019324aa-d057-41bf-a6cd-98fe4feb3e9b)

### History
* :broken_heart: [Build #18198](https://buildkite.com/elastic/integrations/builds/18198) failed 89c0eb002b61c9628b0d10b531775f2212b71084

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18267","number":18267,"commit":"582da2f731faeb5f9e521fd88dd7233fbe8f597a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18198","number":18198,"commit":"89c0eb002b61c9628b0d10b531775f2212b71084"}],"number":18267}
buildkite-pr-comment-->

kcreddy commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,99 @@
+      target_field: blacklens.alert.updated_date
+      formats:
+        - ISO8601
+  - rename:
+      field: json.id
+      target_field: blacklens.alert.id
+      ignore_missing: true
```
copy this also into `event.id`

kcreddy reviewed (commented):
There is a new CI error on static test. You can run it locally to check if it passes: 
`elastic-package test static`. 
From the error, I see the issue is in your `sample_event.json` should only contain the json object [inside your `_source`.](https://github.com/elastic/integrations/blob/582da2f731faeb5f9e521fd88dd7233fbe8f597a/packages/blacklens/data_stream/alerts/sample_event.json#L7-L77) 
[Example](https://github.com/elastic/integrations/blob/main/packages/ti_threatconnect/data_stream/indicator/sample_event.json)

ChriZzn commented on packages/blacklens/manifest.yml:
```diff
@@ -0,0 +1,107 @@
+            multi: false
+            required: true
+            show_user: true
+          - name: tenant_id
+            type: text
+            title: Tenant ID
+            description: Tenant ID of blacklens.io.
+            multi: false
+            required: true
+            show_user: true
```
i added it to httpjson.yml.hbs as Header, this is required to fetch the alerts from API

ChriZzn commented on packages/blacklens/data_stream/alerts/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,99 @@
+  - json:
+      field: event.original
+      target_field: json
+  - fingerprint:
+      fields:
+        - json.id
+        - json.updated_date
+      target_field: _id
+      ignore_missing: true
```
I think this is a bit too complex for the result, the fields that change are primarily only interesting for the UI of blacklens.io, so I'll leave the fingerprinting on json.id for now.

kcreddy replied:
Alright, just want to say that this would mean updated alerts will not be ingested with this fingerprint. 
It would be nice to add this as a `NOTE` in `README` for the users 😄 .

ChriZzn commented:
> There is a new CI error on static test. You can run it locally to check if it passes: `elastic-package test static`. From the error, I see the issue is in your `sample_event.json`. It should only contain the json object [inside your `_source`.](https://github.com/elastic/integrations/blob/582da2f731faeb5f9e521fd88dd7233fbe8f597a/packages/blacklens/data_stream/alerts/sample_event.json#L7-L77) [Example](https://github.com/elastic/integrations/blob/main/packages/ti_threatconnect/data_stream/indicator/sample_event.json)

Hi, i resolved all issues... Regards Christoph

kcreddy commented:
/test

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18320)
* Commit: 591ec352320c584886b52cbf7d6d6613b48b7193

### Failed CI Steps
* [Check integrations blacklens](https://buildkite.com/elastic/integrations/builds/18320#01932a39-a96c-45c3-8e0e-738d405f3e66)

### History
* :broken_heart: [Build #18267](https://buildkite.com/elastic/integrations/builds/18267) failed 582da2f731faeb5f9e521fd88dd7233fbe8f597a
* :broken_heart: [Build #18198](https://buildkite.com/elastic/integrations/builds/18198) failed 89c0eb002b61c9628b0d10b531775f2212b71084

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18320","number":18320,"commit":"591ec352320c584886b52cbf7d6d6613b48b7193"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18267","number":18267,"commit":"582da2f731faeb5f9e521fd88dd7233fbe8f597a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18198","number":18198,"commit":"89c0eb002b61c9628b0d10b531775f2212b71084"}],"number":18320}
buildkite-pr-comment-->

kcreddy commented:
There is a CI error in pipeline tests.
Can you run following tests/commands on your local and if successful, commit all files generated?
First, you have to be at the package root: `cd packages/blacklens`

1. Pipeline Tests (generates `-expected.json` files):
```bash
eval "$(elastic-package stack shellinit)" && elastic-package test pipeline --generate -v
```

2. System Tests (generates `sample_event.json` file):
```bash
eval "$(elastic-package stack shellinit)" && elastic-package test system --generate  -v
```

3. All tests (no files generated, this is to check if all tests are successful):
```bash
eval "$(elastic-package stack shellinit)" && elastic-package test
```

4. Update README, format and lint
```bash
 elastic-package build && elastic-package format && elastic-package lint && elastic-package check && elastic-package build
```

ChriZzn commented:
> There is a CI error in pipeline tests. Can you run following tests/commands on your local and if successful, commit all files generated? First, you have to be at the package root: `cd packages/blacklens`
> 
> 1. Pipeline Tests (generates `-expected.json` files):
> 
> ```shell
> eval "$(elastic-package stack shellinit)" && elastic-package test pipeline --generate -v
> ```
> 
> 2. System Tests (generates `sample_event.json` file):
> 
> ```shell
> eval "$(elastic-package stack shellinit)" && elastic-package test system --generate  -v
> ```
> 
> 3. All tests (no files generated, this is to check if all tests are successful):
> 
> ```shell
> eval "$(elastic-package stack shellinit)" && elastic-package test -v
> ```
> 
> 4. Update README, format and lint
> 
> ```shell
>  elastic-package build && elastic-package format && elastic-package lint && elastic-package check && elastic-package build
> ```

Hi, i ran all tests and did some adjustments... should work now
Regards Christoph

kcreddy commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/18700)
* Commit: 29dc0ff244df0f0ffd087b402cfac7562e87c213

### History
* :broken_heart: [Build #18320](https://buildkite.com/elastic/integrations/builds/18320) failed 591ec352320c584886b52cbf7d6d6613b48b7193
* :broken_heart: [Build #18267](https://buildkite.com/elastic/integrations/builds/18267) failed 582da2f731faeb5f9e521fd88dd7233fbe8f597a
* :broken_heart: [Build #18198](https://buildkite.com/elastic/integrations/builds/18198) failed 89c0eb002b61c9628b0d10b531775f2212b71084

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18700","number":18700,"commit":"29dc0ff244df0f0ffd087b402cfac7562e87c213"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18320","number":18320,"commit":"591ec352320c584886b52cbf7d6d6613b48b7193"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18267","number":18267,"commit":"582da2f731faeb5f9e521fd88dd7233fbe8f597a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/18198","number":18198,"commit":"89c0eb002b61c9628b0d10b531775f2212b71084"}],"number":18700}
buildkite-pr-comment-->

kcreddy reviewed (approved):
LGTM 👍🏼 

> Hi, i ran all tests and did some adjustments... should work now
Regards Christoph

Thanks so much for the contribution 🚀
